Issue 360 popup inside modal

benbcai opened the pull request:
Added the requestFocus and releaseFocus actions in the Modal Manager component to trap and untrap focus in the modal. These actions allow popup components rendered inside the modal the ability to gain focus so that focus is not blocked by the modal. A new example is added in the date picker to render the picker inside the modal.

Resolves #360 

Thanks for contributing to Terra. 
@cerner/terra

[CONTRIBUTORS.md]: ../blob/master/CONTRIBUTORS.md
[License]: ../blob/master/LICENSE

### Deployment links:
https://terra-core-deployed-pr-640.herokuapp.com/static/#/site/datepicker
https://terra-core-deployed-pr-640.herokuapp.com/static/#/tests/date-picker-tests/inside-modal

tbiethman commented on packages/terra-date-picker/src/DateInput.jsx:
```diff
@@ -3,13 +3,18 @@ import PropTypes from 'prop-types';
 import 'terra-base/lib/baseStyles';
 import Button from 'terra-button';
 import IconCalendar from 'terra-icon/lib/icon/IconCalendar';
+import AppDelegate from 'terra-app-delegate';
```
Good stuff Ben! The one thing that I'd change is the inclusion of the AppDelegate at all in the DateInput component; I'd prefer to see the request/release functions passed into this separately.

The AppDelegate has a number of functions that are of no concern to the DateInput component, and I don't want to open the door to any view component having an AppDelegate prop in the future and muddying the waters of what a component's responsibilities are.  

This just means that the component consuming the DateInput (that _is_ receiving an AppDelegate prop) would have to pass in those two functions from the AppDelegate into the DateInput separately.

benbcai replied:
Good point!

benbcai replied:
@tbiethman DatePicker now takes the requestFocus/releaseFocus function instead of the AppDelegate.

https://github.com/cerner/terra-core/pull/640/commits/8eddd676bf0c82b1ecd3750df7ff8e288104aab3

dkasper-was-taken commented on packages/terra-date-picker/src/DateInput.jsx:
```diff
@@ -51,8 +57,42 @@ const defaultProps = {
+  handleOnButtonClick() {
+    // If the containing component (e.g. modal) has the focus trapped, the focus needs to be released so that the picker can get focus.
+    if (this.props.app && this.props.app.releaseFocus) {
+      this.props.app.releaseFocus();
```
Looks to me like the naming seems backward. I would imagine that the picker component would be "requesting" focus for itself from the delegate when it is attempting to display it's FocusTrap, and "releasing" focus when it's hiding it's display.

I would imagine the request/release

benbcai replied:
I was initially thinking of the other way around where the picker component would ask the containing component to "release" the focus. But I see your point for having the picker to "request" focus for itself.

benbcai replied:
@dkasper-was-taken requestFocus now means that popup component is requesting focus and the modal should give up the focus trap. releaseFocus means the popup component no longer needs focus and the modal can regain the focus trap.

https://github.com/cerner/terra-core/pull/640/commits/8eddd676bf0c82b1ecd3750df7ff8e288104aab3

benbcai commented on packages/terra-date-picker/docs/README.md:
```diff
@@ -4,6 +4,8 @@ Terra-date-picker is a controlled input component that provides users a way to e
 
 This is a controlled component because it manages the state of the value in the input. Because this is a controlled input component, it cannot accept the defaultValue prop as it always uses the value prop. React does not allow having both the defaultValue and value props.
 
+The Modal component by default will trap focus. In order to interact with the date picker when rendered inside a modal, the focus on the modal must be released so that the picker can then request focus. The Modal Manager component supports releasing focus on the modal. Therefore, to render the date picker in a modal, it must be composed via the Modal Manager, which will internally handle releasin ...
```
Yeah, I can link the Modal Manager and add some documentation about the modal trapping focus.

benbcai replied:
@JakeLaCombe Add link to ModalManager and added documentation in Modal describing the focus trap.

https://github.com/cerner/terra-core/commit/8eddd676bf0c82b1ecd3750df7ff8e288104aab3

tbiethman commented on packages/terra-date-picker/src/DateInput.jsx:
```diff
@@ -34,23 +34,67 @@ const propTypes = {
 };
 
 const defaultProps = {
+  app: undefined,
```
You can remove this guy now

benbcai replied:
Oops. https://github.com/cerner/terra-core/pull/640/commits/8f373e5a2bb950263491e1e46aeca31473b8ccff

emilyrohrbough commented on packages/terra-app-delegate/tests/jest/AppDelegate.test.jsx:
```diff
@@ -232,13 +262,57 @@ describe('isEqual', () => {
+  });
+
+  it('should return false when releaseFocus does not match', () => {
+    const disclose = () => {};
```
Total nit: could we pull these functions and `appDelegateInstance1` out such that these don't need to be rewritten for each jest test?

benbcai replied:
Updated. https://github.com/cerner/terra-core/pull/640/commits/4ac3c917c262fa149cf0f0d8cc5361dd8e86b48c The number of line in this test got cut in half.

mmalaker commented:
Functional verification completed. The date picker examples display and function correctly within a modal.

mjhenkes commented:
We need to make a couple of changes to the popup before we pull this change in.

benbcai commented:
@dkasper-was-taken I added changes to the Popup component to display inside a modal. Can you take a look at the change? You can just look at the latest commit (### Added
* Added new example to display date picker and popup inside modal.) instead of the entire PR if you like.

@mmalaker I added an example for rendering a popup inside a modal (https://terra-core-deployed-pr-640.herokuapp.com/static/#/site/popup). Could you do some validation/review for the popup inside modal scenario?

mmalaker commented:
Functional verification completed. The popup in modal example displays and functions correctly.
